core/DataAccess/LogAggregator.php

```diff
@@ -373,35 +377,118 @@ public function generateQuery($select, $from, $where, $groupBy, $orderBy, $limit
                         break;
                     }
                 }
-
             }
 
         }
 
         $query = $segment->getSelectQuery($select, $from, $where, $bind, $orderBy, $groupBy, $limit, $offset);
 
+        if (is_array($query) && array_key_exists('sql', $query)) {
+            $query['sql'] = $this->addOriginHintToQuery($query['sql']);
+        }
+
+        return $query;
+    }
+
+    /**
+     * Generate a query / bind array with segment and origin hints for a custom SQL query, similar to generateQuery()
+     * but for use when a query is too complicated to be assembled from SQL fragments and needs to be manually specified
+     *
+     * @param string     $sql       The custom query string
+     * @param array|null $bindOrder Optionally customise the returned bind parameters array by supplying an array
+     *                              which contains a list of standard bind array keys, any order or repetition is
+     *                              supported. eg. [0,1,2,0,1,2] would repeat the standard bind values twice. If omitted
+     *                              then the bind values will be returned once in order: [0,1,2]
+     *
+     * @return array     Array containing the SQL query and bind array, ['sql' => 'SELECT ...', 'bind' => ['abc', 123]]
+     */
+    public function generateQueryCustom(string $sql, ?array $bindOrder = null): array
+    {
+        // Setup bind array
+        $bind = $this->getGeneralQueryBindParams();
+        if ($bindOrder !== null) {
+            $customBind = [];
+            foreach ($bindOrder as $bindKey) {
+                $customBind[] = $bind[$bindKey];
+            }
+            $bind = $customBind;
+        }
+
+        // Apply segment if required.
+        // Ignores enable_segment_cache = 0
+        if (!$this->segment->isEmpty()) {
+
+            $segmentTable = $this->createSegmentTable();
```

Comment:
That method does return the table name without the prefix. Archiving failed locally for me, as I have a table prefix defined. Guess this should fix it:
```suggestion
            $segmentTable = Common::prefixTable($this->createSegmentTable());
```

Reply:
Good spot :+1: `createTemporaryTable()` adds the prefix to the table name itself internally so I've added the prefix in the `createSegmentTable()` return statement after `createTemporaryTable()` has done it's thing.